File: bindings/java/io/dage/Smoke.java
package io.dage;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public final class Smoke {
    public static void main(String[] args) throws Exception {
        Path vectors = args.length == 1
            ? Path.of(args[0])
            : Path.of("tests", "conformance", "bindings");
        String valid = Files.readString(vectors.resolve("valid.json"), StandardCharsets.UTF_8);
        String invalid = Files.readString(vectors.resolve("invalid.json"), StandardCharsets.UTF_8);
        String asyncTool = Files.readString(vectors.resolve("async_tool.json"), StandardCharsets.UTF_8);
        String asyncEffect = Files.readString(vectors.resolve("async_effect.json"), StandardCharsets.UTF_8);
        ExecutorService runExecutor = Executors.newSingleThreadExecutor();
        ExecutorService callbackExecutor = Executors.newSingleThreadExecutor();
        try (Engine engine = new Engine()) {
            engine.registerAsyncExecutor(
                "java_async_echo",
                (context, input) -> CompletableFuture.completedFuture(input),
                callbackExecutor);
            engine.registerAsyncExecutor(
                "java_async_fail",
                (context, input) -> CompletableFuture.failedFuture(
                    new IllegalStateException("expected Java Executor failure")),
                callbackExecutor);
            engine.registerAsyncExecutor("java_async_effect", (context, input) -> {
                if (context.idempotencyKey().isEmpty()) {
                    return CompletableFuture.failedFuture(
                        new AssertionError("missing idempotency key"));
                }
                context.commitEffect();
                return CompletableFuture.completedFuture(input);
            }, callbackExecutor);
            try (Workflow workflow = engine.load(valid)) {
                String graph = workflow.mermaid();
                if (!graph.contains("flowchart TD") || !graph.contains("done-完成")) {
                    throw new AssertionError("unexpected Mermaid output");
                }
                try (Run run = engine.createRun(workflow)) {
                    String output = run.executeAsync("{}", runExecutor).get(5, TimeUnit.SECONDS);
                    if (!output.contains("\"conformance\":true")
                            || !output.contains("基础设施-🚀")) {
                        throw new AssertionError("unexpected Run output");
                    }
                }
                Run ownershipProbe = engine.createRun(workflow);
                try {
                    assertEngineRejectsLiveRun(engine);
                } finally {
                    ownershipProbe.close();
                }
                try (Run run = engine.createRun(workflow)) {
                    run.cancel("Java conformance pre-cancel");
                    try {
                        run.executeAsync("{}", runExecutor).get(5, TimeUnit.SECONDS);
                        throw new AssertionError("cancelled Run completed successfully");
                    } catch (ExecutionException expected) {
                        if (!(expected.getCause() instanceof DageException)) throw expected;
                    }
                }
            }
            exerciseAsyncExecutors(engine, runExecutor, callbackExecutor, asyncTool, asyncEffect);
            try {
                engine.load(invalid).close();
                throw new AssertionError("invalid Workflow was accepted");
            } catch (DageException expected) {
                // The binding must translate native validation failure.
            }
        } finally {
            runExecutor.shutdownNow();
            callbackExecutor.shutdownNow();
            if (!runExecutor.awaitTermination(5, TimeUnit.SECONDS)
                    || !callbackExecutor.awaitTermination(5, TimeUnit.SECONDS))
                throw new AssertionError("Java conformance executor did not terminate");
        }
        System.out.println("Java binding conformance passed");
    }

    private static void assertEngineRejectsLiveRun(Engine engine) {
        try {
            engine.close();
            throw new AssertionError("Engine closed with a live Run");
        } catch (IllegalStateException expected) {
            // C Run error handling retains the Engine handle.
        }
    }

    private static void exerciseAsyncExecutors(
            Engine engine, ExecutorService runExecutor, ExecutorService callbackExecutor,
            String asyncTool, String asyncEffect)
            throws Exception {
        try (Workflow workflow = engine.load(asyncTool.replace("__EXECUTOR__", "java_async_echo"));
                Run run = engine.createRun(workflow)) {
            String output = run.executeAsync("{\"message\":\"Java-🚀\"}", runExecutor)
                .get(5, TimeUnit.SECONDS);
            if (!output.contains("Java-🚀")) throw new AssertionError("async output lost UTF-8");
        }
        try (Workflow workflow = engine.load(asyncTool.replace("__EXECUTOR__", "java_async_fail"));
                Run run = engine.createRun(workflow)) {
            try {
                run.executeAsync("{}", runExecutor).get(5, TimeUnit.SECONDS);
                throw new AssertionError("async Executor exception was not translated");
            } catch (ExecutionException expected) {
                if (!(expected.getCause() instanceof DageException)) throw expected;
            }
        }
        String effectWorkflow = asyncEffect.replace("__EXECUTOR__", "java_async_effect");
        try (Workflow workflow = engine.load(effectWorkflow);
                Run run = engine.createRun(workflow)) {
            String output = run.executeAsync("{\"effect\":true}", runExecutor)
                .get(5, TimeUnit.SECONDS);
            if (!output.contains("\"effect\":true")) {
                throw new AssertionError("effect completion failed");
            }
        }

        CountDownLatch started = new CountDownLatch(1);
        CountDownLatch cancellationSeen = new CountDownLatch(1);
        engine.registerAsyncExecutor("java_async_wait", (context, input) -> {
            CompletableFuture<String> result = new CompletableFuture<>();
            started.countDown();
            callbackExecutor.execute(() -> {
                while (!context.isCancelled()) Thread.onSpinWait();
                cancellationSeen.countDown();
                result.complete(input);
            });
            return result;
        }, callbackExecutor);
        try (Workflow workflow = engine.load(asyncTool.replace("__EXECUTOR__", "java_async_wait"));
                Run run = engine.createRun(workflow)) {
            CompletableFuture<String> future = run.executeAsync("{}", runExecutor);
            if (!started.await(5, TimeUnit.SECONDS)) {
                throw new AssertionError("async Executor did not start");
            }
            if (!future.cancel(false)) throw new AssertionError("Run cancellation was rejected");
            if (!cancellationSeen.await(5, TimeUnit.SECONDS)) {
                throw new AssertionError("async Executor did not observe cancellation");
            }
        }
    }
}

File: bindings/rust/tests/smoke.rs
use dage::{Engine, Error, RunFuture, ThreadSpawner};
use std::fs;
use std::future::Future;
use std::path::PathBuf;
use std::pin::Pin;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::task::{Context, Poll, Waker};
use std::time::{Duration, Instant};

fn wait(mut future: RunFuture) -> dage::Result<String> {
    let mut context = Context::from_waker(Waker::noop());
    let deadline = Instant::now() + Duration::from_secs(5);
    loop {
        if let Poll::Ready(value) = Pin::new(&mut future).poll(&mut context) {
            return value;
        }
        assert!(Instant::now() < deadline, "Rust RunFuture timed out");
        std::thread::sleep(Duration::from_millis(1));
    }
}

#[test]
fn shared_binding_conformance() {
    let vectors = std::env::var_os("DAGE_CONFORMANCE_DIR")
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("../../tests/conformance/bindings"));
    let valid = fs::read_to_string(vectors.join("valid.json")).unwrap();
    let invalid = fs::read_to_string(vectors.join("invalid.json")).unwrap();
    let async_tool = fs::read_to_string(vectors.join("async_tool.json")).unwrap();
    let async_effect = fs::read_to_string(vectors.join("async_effect.json")).unwrap();
    let engine = Engine::new().unwrap();
    let workflow = engine.load(&valid).unwrap();
    let graph = workflow.mermaid().unwrap();
    assert!(graph.contains("flowchart TD") && graph.contains("done-完成"));
    assert!(engine.load(&invalid).is_err());

    let spawner = Arc::new(ThreadSpawner);
    engine
        .register_async_executor(
            "rust_echo",
            |_context, input| Box::pin(async move { Ok(input) }),
            spawner.clone(),
        )
        .unwrap();
    engine
        .register_async_executor(
            "rust_fail",
            |_context, _input| Box::pin(async { Err(Error(5)) }),
            spawner.clone(),
        )
        .unwrap();
    engine
        .register_async_executor(
            "rust_effect",
            |context, input| {
                Box::pin(async move {
                    assert!(!context.idempotency_key.is_empty());
                    context.commit_effect()?;
                    Ok(input)
                })
            },
            spawner.clone(),
        )
        .unwrap();

    let workflow_json = |executor: &str| async_tool.replace("__EXECUTOR__", executor);
    let workflow = engine.load(&workflow_json("rust_echo")).unwrap();
    let run = engine.create_run(&workflow).unwrap();
    let output = wait(
        run.execute_async("{\"message\":\"Rust-🚀\"}".into(), spawner.clone())
            .unwrap(),
    )
    .unwrap();
    assert!(output.contains("Rust-🚀"));

    let workflow = engine.load(&workflow_json("rust_fail")).unwrap();
    let run = engine.create_run(&workflow).unwrap();
    assert_eq!(
        wait(run.execute_async("{}".into(), spawner.clone()).unwrap()),
        Err(Error(5))
    );

    let effect = async_effect.replace("__EXECUTOR__", "rust_effect");
    let workflow = engine.load(&effect).unwrap();
    let run = engine.create_run(&workflow).unwrap();
    assert!(wait(
        run.execute_async("{\"effect\":true}".into(), spawner.clone())
            .unwrap()
    )
    .unwrap()
    .contains("\"effect\":true"));

    let started = Arc::new(AtomicBool::new(false));
    let seen = Arc::new(AtomicBool::new(false));
    let started_callback = started.clone();
    let seen_callback = seen.clone();
    engine
        .register_async_executor(
            "rust_wait",
            move |context, input| {
                let started = started_callback.clone();
                let seen = seen_callback.clone();
                Box::pin(async move {
                    started.store(true, Ordering::Release);
                    while !context.is_cancelled() {
                        std::thread::yield_now();
                    }
                    seen.store(true, Ordering::Release);
                    Ok(input)
                })
            },
            spawner.clone(),
        )
        .unwrap();
    let workflow = engine.load(&workflow_json("rust_wait")).unwrap();
    let run = engine.create_run(&workflow).unwrap();
    let future = run.execute_async("{}".into(), spawner).unwrap();
    let deadline = Instant::now() + Duration::from_secs(5);
    while !started.load(Ordering::Acquire) {
        assert!(
            Instant::now() < deadline,
            "Rust async Executor did not start"
        );
        std::thread::yield_now();
    }
    drop(future);
    while !seen.load(Ordering::Acquire) {
        assert!(
            Instant::now() < deadline,
            "Rust async Executor did not observe cancellation"
        );
        std::thread::yield_now();
    }
}
